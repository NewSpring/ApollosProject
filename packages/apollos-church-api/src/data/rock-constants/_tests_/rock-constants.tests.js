import { fetch } from 'apollo-server-env';
import ApollosConfig from '@apollosproject/config';
import { dataSource as RockConstants } from '../index';

ApollosConfig.loadJs({
  ROCK_MAPPINGS: {
    PRAYER_REQUEST_TYPE: 'PrayerRequest',
    INTERACTIONS: {
      CHANNEL_NAME: 'Apollos App',
      COMPONENT_NAME: 'Apollos Prayer Request',
      CHANNEL_MEDIUM_TYPE_ID: 512,
    },
  },
});

const buildGetMock = (response, dataSource) => {
  const get = jest.fn();
  if (Array.isArray(response) && Array.isArray(response[0])) {
    response.forEach((responseVal) => {
      get.mockReturnValueOnce(
        new Promise((resolve) => resolve(dataSource.normalize(responseVal)))
      );
    });
  }
  get.mockReturnValue(
    new Promise((resolve) => resolve(dataSource.normalize(response)))
  );
  return get;
};

describe('RockConstants', () => {
  beforeEach(() => {
    fetch.resetMocks();
  });
  it("creates a Channel if it doesn't exist", async () => {
    const dataSource = new RockConstants();
    dataSource.prayerModelType = buildGetMock({ id: 101 }, dataSource);
    dataSource.get = buildGetMock([[], { id: 1 }], dataSource);
    dataSource.post = buildGetMock('1', dataSource);
    const result = await dataSource.prayerRequestInteractionChannel();
    expect(result).toMatchSnapshot();
    expect(dataSource.prayerModelType.mock.calls).toMatchSnapshot();
    expect(dataSource.get.mock.calls).toMatchSnapshot();
    expect(dataSource.post.mock.calls).toMatchSnapshot();
  });
  it('finds the Channel if it exists', async () => {
    const dataSource = new RockConstants();
    dataSource.get = buildGetMock([{ Id: 1 }], dataSource);
    dataSource.post = jest.fn();
    const result = await dataSource.prayerRequestInteractionChannel();
    expect(result).toMatchSnapshot();
    expect(dataSource.get.mock.calls).toMatchSnapshot();
    expect(dataSource.post.mock.calls.length).toBe(0);
  });
  it("creates a Component if it doesn't exist", async () => {
    const dataSource = new RockConstants();
    dataSource.prayerModelType = buildGetMock({ id: 101 }, dataSource);
    dataSource.get = buildGetMock([[], { id: 1 }], dataSource);
    dataSource.post = buildGetMock('1', dataSource);
    const result = await dataSource.prayerRequestInteractionComponent({
      prayerRequestId: 7,
      prayerName: 'Some Title',
    });
    expect(result).toMatchSnapshot();
    expect(dataSource.prayerModelType.mock.calls).toMatchSnapshot();
    expect(dataSource.get.mock.calls).toMatchSnapshot();
    expect(dataSource.post.mock.calls).toMatchSnapshot();
  });
  it('finds the Component if it exists', async () => {
    const dataSource = new RockConstants();
    dataSource.get = buildGetMock([{ Id: 1 }], dataSource);
    dataSource.post = jest.fn();
    const result = await dataSource.prayerRequestInteractionComponent({
      prayerRequestId: 7,
      prayerName: 'Some Title',
    });
    expect(result).toMatchSnapshot();
    expect(dataSource.get.mock.calls).toMatchSnapshot();
    expect(dataSource.post.mock.calls.length).toBe(0);
  });
  it('finds a Prayer Request model ID', async () => {
    const dataSource = new RockConstants();
    dataSource.get = buildGetMock([{ Id: 1 }], dataSource);
    const result = await dataSource.prayerModelType('PrayerRequest');
    expect(result).toMatchSnapshot();
    expect(dataSource.get.mock.calls).toMatchSnapshot();
  });
  it('Throws when finding an unknown model ', () => {
    const dataSource = new RockConstants();
    dataSource.get = buildGetMock([{ Id: 1 }], dataSource);
    const prom = dataSource.prayerModelType('IDontExist');
    expect(prom).rejects.toEqual(
      new Error('IDontExist has not been mapped into a Rock type!')
    );
  });
  it('Returns null if model type not found ', async () => {
    const dataSource = new RockConstants();
    dataSource.get = buildGetMock([], dataSource);
    const result = await dataSource.prayerModelType('PrayerRequest');
    expect(result).toEqual(null);
  });
});
