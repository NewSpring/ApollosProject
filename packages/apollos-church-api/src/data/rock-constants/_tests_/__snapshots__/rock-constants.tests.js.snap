// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RockConstants creates a Channel if it doesn't exist 1`] = `
Object {
  "id": 1,
}
`;

exports[`RockConstants creates a Channel if it doesn't exist 2`] = `
Array [
  Array [
    "PrayerRequest",
  ],
]
`;

exports[`RockConstants creates a Channel if it doesn't exist 3`] = `
Array [
  Array [
    "InteractionChannels?%24filter=%28Name%20eq%20%27Apollos%20App%27%29%20and%20%28UsesSession%20eq%20true%29%20and%20%28IsActive%20eq%20true%29%20and%20%28ComponentEntityTypeId%20eq%20101%29%20and%20%28ChannelTypeMediumValueId%20eq%20512%29",
    Object {},
    Object {
      "ttl": 86400,
    },
  ],
  Array [
    "/InteractionChannels/1",
  ],
]
`;

exports[`RockConstants creates a Channel if it doesn't exist 4`] = `
Array [
  Array [
    "/InteractionChannels",
    Object {
      "ChannelTypeMediumValueId": 512,
      "ComponentEntityTypeId": 101,
      "IsActive": true,
      "Name": "Apollos App",
      "UsesSession": true,
    },
  ],
]
`;

exports[`RockConstants creates a Component if it doesn't exist 1`] = `Array []`;

exports[`RockConstants creates a Component if it doesn't exist 2`] = `
Array [
  Array [
    "PrayerRequest",
  ],
]
`;

exports[`RockConstants creates a Component if it doesn't exist 3`] = `
Array [
  Array [
    "InteractionChannels?%24filter=%28Name%20eq%20%27Apollos%20App%27%29%20and%20%28UsesSession%20eq%20true%29%20and%20%28IsActive%20eq%20true%29%20and%20%28ComponentEntityTypeId%20eq%20101%29%20and%20%28ChannelTypeMediumValueId%20eq%20512%29",
    Object {},
    Object {
      "ttl": 86400,
    },
  ],
  Array [
    "/InteractionChannels/1",
  ],
  Array [
    "InteractionComponents?%24filter=%28Name%20eq%20%27undefined%20-%20Some%20Title%27%29%20and%20%28ChannelId%20eq%201%29%20and%20%28EntityId%20eq%207%29",
    Object {},
    Object {
      "ttl": 86400,
    },
  ],
]
`;

exports[`RockConstants creates a Component if it doesn't exist 4`] = `
Array [
  Array [
    "/InteractionChannels",
    Object {
      "ChannelTypeMediumValueId": 512,
      "ComponentEntityTypeId": 101,
      "IsActive": true,
      "Name": "Apollos App",
      "UsesSession": true,
    },
  ],
]
`;

exports[`RockConstants finds a Prayer Request model ID 1`] = `
Object {
  "id": 1,
}
`;

exports[`RockConstants finds a Prayer Request model ID 2`] = `
Array [
  Array [
    "EntityTypes?%24filter=Name%20eq%20%27Rock.Model.PrayerRequest%27",
    Object {},
    Object {
      "ttl": 86400,
    },
  ],
]
`;

exports[`RockConstants finds the Channel if it exists 1`] = `
Object {
  "id": 1,
}
`;

exports[`RockConstants finds the Channel if it exists 2`] = `
Array [
  Array [
    "EntityTypes?%24filter=Name%20eq%20%27Rock.Model.PrayerRequest%27",
    Object {},
    Object {
      "ttl": 86400,
    },
  ],
  Array [
    "InteractionChannels?%24filter=%28Name%20eq%20%27Apollos%20App%27%29%20and%20%28UsesSession%20eq%20true%29%20and%20%28IsActive%20eq%20true%29%20and%20%28ComponentEntityTypeId%20eq%201%29%20and%20%28ChannelTypeMediumValueId%20eq%20512%29",
    Object {},
    Object {
      "ttl": 86400,
    },
  ],
]
`;

exports[`RockConstants finds the Component if it exists 1`] = `
Object {
  "id": 1,
}
`;

exports[`RockConstants finds the Component if it exists 2`] = `
Array [
  Array [
    "EntityTypes?%24filter=Name%20eq%20%27Rock.Model.PrayerRequest%27",
    Object {},
    Object {
      "ttl": 86400,
    },
  ],
  Array [
    "InteractionChannels?%24filter=%28Name%20eq%20%27Apollos%20App%27%29%20and%20%28UsesSession%20eq%20true%29%20and%20%28IsActive%20eq%20true%29%20and%20%28ComponentEntityTypeId%20eq%201%29%20and%20%28ChannelTypeMediumValueId%20eq%20512%29",
    Object {},
    Object {
      "ttl": 86400,
    },
  ],
  Array [
    "InteractionComponents?%24filter=%28Name%20eq%20%27undefined%20-%20Some%20Title%27%29%20and%20%28ChannelId%20eq%201%29%20and%20%28EntityId%20eq%207%29",
    Object {},
    Object {
      "ttl": 86400,
    },
  ],
]
`;
